### PR TITLE
fpng: add recipe

### PR DESCRIPTION
**Pull request by toge**
Specify library name and version:  **fpng/***

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/fpng/all/test_package/test_package.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#include <vector>
+#include <cstdint>
+
+#include "fpng.h"
+
+int main() {
+	fpng::fpng_init();
```
Comment (valgur) [inline recipes/fpng/all/test_package/test_package.cpp:7]: Inconsistent indentation.

File: recipes/fpng/all/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,30 @@
+cmake_minimum_required(VERSION 3.8)
```
Comment (valgur) [inline recipes/fpng/all/CMakeLists.txt:1]:
```suggestion
cmake_minimum_required(VERSION 3.15)
```
3.15 is required by Conan v2 anyway.

File: recipes/fpng/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    settings = "os", "arch", "compiler", "build_type"
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+        "with_sse": [True, False],
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+        "with_sse": False,
```
Comment (valgur) [inline recipes/fpng/all/conanfile.py:27]:
```suggestion
        "with_sse": True,
```
According to https://store.steampowered.com/hwsurvey/ (other settings), SSE4.1 support is at 99.66% in consumer-grade hardware. I would enable it by default.

**Review by valgur — APPROVED** [review]

File: recipes/fpng/all/test_package/test_package.cpp (hunk 1 above)
Comment (toge) [reply recipes/fpng/all/test_package/test_package.cpp]: @valgur 
Thanks a lot!
Fixed.

**Review by valgur — APPROVED** [review]

File: recipes/fpng/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+from conan.tools.build import check_min_cppstd
+import os
+
+required_conan_version = ">=1.53.0"
+
+
+class FpngConan(ConanFile):
+    name = "fpng"
+    description = "Super fast C++ .PNG writer/reader"
+    license = "unlicense",
```
Comment (valgur) [inline recipes/fpng/all/conanfile.py:13]:
```suggestion
    license = "Unlicense",
```
Use the conventional capitalization for the SPDX identifier. Probably does not matter, though.

File: recipes/fpng/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    def validate(self):
+        if self.settings.compiler.get_safe("cppstd"):
+            check_min_cppstd(self, self._min_cppstd)
+
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["FPNG_SRC_DIR"] = self.source_folder.replace("\\", "/")
```
Comment (valgur) [inline recipes/fpng/all/conanfile.py:55]: Not worth changing it here, but I generally find it easier to
```py
def export_sources(self):
    copy(self, "CMakeLists.txt", self.recipe_folder, os.path.join(self.export_sources_folder, "src"))
```
That way all of the CMakeLists.txt and source paths work without any prefixing.

The current approach is not wrong either, though, of course.

File: recipes/fpng/all/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,30 @@
+cmake_minimum_required(VERSION 3.15)
+project(fpng LANGUAGES CXX)
+
+include(GNUInstallDirs)
+
+add_library(fpng ${FPNG_SRC_DIR}/src/fpng.cpp)
+set_target_properties(fpng PROPERTIES
+    PUBLIC_HEADER ${FPNG_SRC_DIR}/src/fpng.h
+    WINDOWS_EXPORT_ALL_SYMBOLS ON
```
Comment (valgur) [inline recipes/fpng/all/CMakeLists.txt:9]: I don't see the project build files doing the equivalent, so the recipe should probably simply disable the Windows shared configuration instead.

Comment (AbrilRBS) [reply recipes/fpng/all/CMakeLists.txt:9]: +1 @toge

File: recipes/fpng/all/conanfile.py (hunk 5 above)
Comment (toge) [reply recipes/fpng/all/conanfile.py:55]: @valgur 
Sorry for my late response.

I tried your comment, but I failed.
Are there any misunderstandings in my modification?

Error message is following.

```
fpng/1.0.6: Running CMake.configure()
fpng/1.0.6: RUN: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/toge/.conan2/p/b/fpng8db6eab8b7baf/p" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/home/toge/.conan2/p/b/fpng8db6eab8b7baf/b/src"
CMake Warning:
  Ignoring extra path from command line:

   "/home/toge/.conan2/p/b/fpng8db6eab8b7baf/b/src"


CMake Error: The source directory "/home/toge/.conan2/p/b/fpng8db6eab8b7baf/b/src" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
```

modified conanfile.py
<details>
```
from conan import ConanFile
from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
from conan.tools.files import get, save, load, copy, rm
from conan.tools.build import check_min_cppstd
import os

required_conan_version = ">=1.53.0"


class FpngConan(ConanFile):
    name = "fpng"
    description = "Super fast C++ .PNG writer/reader"
    license = "Unlicense",
    url = "https://github.com/conan-io/conan-center-index"
    homepage = "https://github.com/richgel999/fpng"
    topics = ("png", "writer", "reader")
    package_type = "library"
    settings = "os", "arch", "compiler", "build_type"
    options = {
        "shared": [True, False],
        "fPIC": [True, False],
        "with_sse": [True, False],
    }
    default_options = {
        "shared": False,
        "fPIC": True,
        "with_sse": True,
    }

    def export_sources(self):
        copy(self, "CMakeLists.txt", self.recipe_folder, os.path.join(self.export_sources_folder, "src"))

    def config_options(self):
        if self.settings.os == "Windows":
            del self.options.fPIC

    def configure(self):
        if self.options.shared:
            self.options.rm_safe("fPIC")

    def layout(self):
        cmake_layout(self, src_folder="src")

    def validate(self):
        check_min_cppstd(self, 11)

    def source(self):
        get(self, **self.conan_data["sources"][self.version], strip_root=True)
        # remove original CMakeLists.txt (which builds a test program only)
        rm(self, "CMakeLists.txt", self.source_folder)

    def generate(self):
        tc = CMakeToolchain(self)
        tc.variables["FPNG_WITH_SSE"] = self.options.with_sse
        tc.generate()

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()

    def package(self):
        filename = os.path.join(self.source_folder, "src", "fpng.cpp")
        file_content = load(self, filename)
        license_start = "	This is free and unencumbered software"
        license_end = "*/"
        license_contents = file_content[file_content.find(license_start):file_content.find(license_end)]
        save(self, os.path.join(self.package_folder, "licenses", "LICENSE"), license_contents)
        cmake = CMake(self)
        cmake.install()

    def package_info(self):
        self.cpp_info.libs = ["fpng"]
        if self.settings.os in ("FreeBSD", "Linux"):
            self.cpp_info.system_libs = ["m"]
```
</details>

**Comment by toge** [comment]
close due to limit of number of open PR.

**Pull request closed without merging**